src/engine/controls/macrocontrol.cpp

```diff
@@ -0,0 +1,234 @@
+#include "macrocontrol.h"
+
+namespace {
+constexpr uint kRecordingTimerInterval = 100;
+constexpr size_t kRecordingQueueSize = kRecordingTimerInterval / 10;
+} // namespace
+
+ConfigKey MacroControl::getConfigKey(QString name) {
+    return ConfigKey(m_group, m_controlPattern.arg(name));
+}
+
+MacroControl::MacroControl(QString group, UserSettingsPointer pConfig, int slot)
+        : EngineControl(group, pConfig),
+          m_slot(slot),
+          m_controlPattern(QString("macro_%1_%2").arg(slot)),
+          m_bJumpPending(false),
+          m_recordedActions(kRecordingQueueSize),
+          m_iNextAction(0),
+          m_COStatus(getConfigKey("status")),
+          m_COIndicator(getConfigKey("indicator")),
+          m_record(getConfigKey("record")),
+          m_toggle(getConfigKey("toggle")),
+          m_clear(getConfigKey("clear")),
+          m_activate(getConfigKey("activate")) {
+    m_COIndicator.setReadOnly();
+    m_COStatus.setReadOnly();
+    setStatus(Status::NoTrack);
+
+    m_updateRecordingTimer.moveToThread(qApp->thread());
+    connect(&m_updateRecordingTimer,
+            &QTimer::timeout,
+            this,
+            &MacroControl::updateRecording);
+
+    m_record.setButtonMode(ControlPushButton::TRIGGER);
+    connect(&m_record,
+            &ControlObject::valueChanged,
+            this,
+            &MacroControl::controlRecord);
+    m_toggle.setButtonMode(ControlPushButton::TRIGGER);
+    connect(&m_toggle,
+            &ControlObject::valueChanged,
+            this,
+            &MacroControl::controlToggle);
+    m_clear.setButtonMode(ControlPushButton::TRIGGER);
+    connect(&m_clear,
+            &ControlObject::valueChanged,
+            this,
+            &MacroControl::controlClear);
+    m_activate.setButtonMode(ControlPushButton::TRIGGER);
+    connect(&m_activate,
+            &ControlObject::valueChanged,
+            this,
+            &MacroControl::controlActivate);
+}
+
+void MacroControl::process(const double dRate, const double dCurrentSample, const int iBufferSize) {
+    Q_UNUSED(dRate);
+    m_bJumpPending = false;
+    if (!isPlaying()) {
+        return;
+    }
+    double framePos = dCurrentSample / mixxx::kEngineChannelCount;
+    const MacroAction& nextAction = m_pMacro->getActions().at(m_iNextAction);
```

Comment:
well, the Macro would have to be _stopped_ and _cleared_ in the same breath, I am not sure whether this is a legitimate concern. And if it is, we would have to use a lock, because that could in theory happen almost everywhere.